Implement a fallback in CompactPtr for storing some pointers that don't fit within 36-bits.

MenloDorian opened the pull request:
#### 4d03df5acb1fcbcab2a05f510e956d3cb770d37a
<pre>
Implement a fallback in CompactPtr for storing some pointers that don&apos;t fit within 36-bits.
<a href="https://bugs.webkit.org/show_bug.cgi?id=303343">https://bugs.webkit.org/show_bug.cgi?id=303343</a>
<a href="https://rdar.apple.com/165647792">rdar://165647792</a>

Reviewed by Dan Hecht.

The OS linker / loader (at its own discretion) may choose to locate statically allocated
objects at addresses that are beyond the 36-bits range (unlike heap addresses which are
always within 36-bits).  As such, for robustness, CompactPtr should have a fallback system
that allows it to encode such outsized pointers.

We observe that:
1. The OS will never allocate objects (heap or otherwise) within the __PAGEZERO region.
2. The number of such statically allocated objects that we&apos;ll ever store in CompactPtrs
   are finite and small-ish (on the order of &lt; 1100 instances).

Hence, we can use the addresses within __PAGEZERO to represent indexes into a table of
OutsizedCompactPtrs where the full (&gt; 36 bits) pointer value is actually stored.

__PAGEZERO is currently around 4G in size.  However, we&apos;ll conservatively reserve only
the first 256K of addresses for OutsizedCompactPtrs. This allows us to encode up to 16K
outsized pointers.

Meanwhile, we should also reduce the number of statically allocated objects that can be
stored in CompactPtrs.  It would be ideal if the number of such objects reduce to way
under 1022. With that, we would be able to encode all those pointers even if the size of
__PAGEZERO is literally reduced to the size of 1 16K page.  This will be left as an
exercise for future patches.  Until then, we&apos;ll work with the 256K heuristic.

Test: Tools/TestWebKitAPI/Tests/WTF/CompactPtr.cpp
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/CompactPtr.cpp: Added.
(WTF::ensureOutsizedCompactPtrManager):
(WTF::OutsizedCompactPtr::encode):
(WTF::OutsizedCompactPtr::decode):
* Source/WTF/wtf/CompactPtr.h:
(WTF::CompactPtr::encode):
(WTF::CompactPtr::decode):
* Tools/TestWebKitAPI/Tests/WTF/CompactPtr.cpp:
(TestWebKitAPI::TEST(WTF_CompactPtr, Basic)):

Canonical link: <a href="https://commits.webkit.org/303809@main">https://commits.webkit.org/303809@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/210c65b32493db13887e4831a0748da480f4e3cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133616 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6121 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44778 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141189 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85670 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135486 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6642 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5985 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102220 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d2233284-e92b-4eef-beca-32f163ea6fe6) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136563 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4704 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119791 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83020 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8f3ee8d6-85e8-469e-9fbc-61f1cab53e72) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4583 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2192 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125693 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113719 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37905 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143843 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132130 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5793 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38486 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110603 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5873 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4963 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110787 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4438 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116044 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59542 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20662 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5845 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34355 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165093 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5693 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69303 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43136 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5953 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5800 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/WTF/wtf/CompactPtr.h:
```diff
@@ -179,7 +225,11 @@ class CompactPtr {
+        StorageType candidate = intPtr >> bitsShift;
+        if ((static_cast<uintptr_t>(candidate) << bitsShift) != intPtr) [[unlikely]]
+            return std::bit_cast<StorageType>(OutsizedCompactPtr::encode(std::bit_cast<void*>(ptr)));
+        StorageType encoded = static_cast<StorageType>(candidate);
```
Cast unnecessary, `candidate` is a `StorageType`

MenloDorian replied:
Fixed.  I renamed `candidate` to `encoded` as weak as remove the now redundant old `encoded` and its also redundant ASSERT.

dhecht commented on Source/WTF/wtf/CompactPtr.cpp:
```diff
@@ -0,0 +1,88 @@
+    SegmentedVector<void*> outsizedPointers;
+};
+
+static Lock outsizedComapctPtrLock;
```
fix spelling `Comapct`

dhecht commented on Tools/TestWebKitAPI/Tests/WTF/CompactPtr.cpp:
```diff
@@ -50,6 +51,15 @@ TEST(WTF_CompactPtr, Basic)
+        CompactPtr<AlignedRefLogger> sPtr(&staticA);
+        EXPECT_EQ(&staticA, sPtr.get());
+        EXPECT_EQ(&staticA, &*sPtr);
+        EXPECT_EQ(&staticA.name, &sPtr->name);
```
maybe have the test case add a second entry before decoding the first to prove that they occupy distinct table entries.

dhecht replied:
Oh and maybe encode the same pointer twice and verify that they occupy the same entry?

dhecht reviewed (approved):
r=me with comments

Commit:
Implement a fallback in CompactPtr for storing some pointers that don't fit within 36-bits.
https://bugs.webkit.org/show_bug.cgi?id=303343
rdar://165647792

Reviewed by Dan Hecht.

The OS linker / loader (at its own discretion) may choose to locate statically allocated
objects at addresses that are beyond the 36-bits range (unlike heap addresses which are
always within 36-bits).  As such, for robustness, CompactPtr should have a fallback system
that allows it to encode such outsized pointers.

We observe that:
1. The OS will never allocate objects (heap or otherwise) within the __PAGEZERO region.
2. The number of such statically allocated objects that we'll ever store in CompactPtrs
   are finite and small-ish (on the order of < 1100 instances).

Hence, we can use the addresses within __PAGEZERO to represent indexes into a table of
OutsizedCompactPtrs where the full (> 36 bits) pointer value is actually stored.

__PAGEZERO is currently around 4G in size.  However, we'll conservatively reserve only
the first 256K of addresses for OutsizedCompactPtrs. This allows us to encode up to 16K
outsized pointers.

Meanwhile, we should also reduce the number of statically allocated objects that can be
stored in CompactPtrs.  It would be ideal if the number of such objects reduce to way
under 1022. With that, we would be able to encode all those pointers even if the size of
__PAGEZERO is literally reduced to the size of 1 16K page.  This will be left as an
exercise for future patches.  Until then, we'll work with the 256K heuristic.

Test: Tools/TestWebKitAPI/Tests/WTF/CompactPtr.cpp
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/CompactPtr.cpp: Added.
(WTF::ensureOutsizedCompactPtrManager):
(WTF::OutsizedCompactPtr::encode):
(WTF::OutsizedCompactPtr::decode):
* Source/WTF/wtf/CompactPtr.h:
(WTF::CompactPtr::encode):
(WTF::CompactPtr::decode):
* Tools/TestWebKitAPI/Tests/WTF/CompactPtr.cpp:
(TestWebKitAPI::TEST(WTF_CompactPtr, Basic)):

Canonical link: https://commits.webkit.org/303809@main
